xtractmime/_utils.py

```diff
@@ -0,0 +1,272 @@
+            index += number_size
+
+            if index >= input_size - 4:
+                break
+
+            if input_bytes[index : index + 4] == b"webm":  # noqa: E203
+                return True
+        index += 1
+
+    return False
+
+
+def match_mp3_header(input_bytes: bytes, input_size: int, index: int) -> bool:
+    if input_size < 4:
+        return False
+
+    if (
+        input_bytes[index : index + 1] != b"\xff"  # noqa: E203
+        or bytes([input_bytes[index + 1] & 224]) != b"\xe0"
+    ):
+        return False
+
+    layer = (input_bytes[index + 1] & 6) >> 1
+    if layer == 0:
+        return False
+
+    bit_rate = (input_bytes[index + 2] & 240) >> 4
+
+    if bit_rate == 15:
+        return False
+
+    sample_rate = (input_bytes[index + 2] & 12) >> 2
+    if sample_rate == 3:
+        return False
+
+    final_layer = (input_bytes[index + 1] & 6) >> 1
+
+    if 4 - final_layer != 3:
+        return False
+
+    return True
+
+
+def parse_mp3_frame(input_bytes: bytes) -> Tuple[int, int, int, int]:
+    version = (input_bytes[1] & 24) >> 3
+    bit_rate_index = (input_bytes[2] & 240) >> 4
+
+    if version & 1:
+        bit_rate = MP3_RATES[bit_rate_index]
+    else:
+        bit_rate = MP25_RATES[bit_rate_index]
+
+    sample_rate_index = (input_bytes[2] & 12) >> 2
+    freq = SAMPLE_RATES[sample_rate_index]
+
+    if version == 2:
+        freq >>= 1
+    elif version == 0:
+        freq >>= 2
+
+    pad = (input_bytes[2] & 2) >> 1
+
+    return version, bit_rate, freq, pad
+
+
+def mp3_framesize(version, bit_rate, freq, pad) -> int:
+    if (version & 1) == 0:
+        scale = 72
+    else:
+        scale = 144
+
+    size = bit_rate * scale / freq
+
+    if pad:
+        size += 1
+
+    return int(size)
+
+
+def is_mp3_non_ID3_signature(input_bytes: bytes) -> bool:
+    """Return True if the input bytes belong to an MP3 file without ID3
+    metadata, or False otherwise.
+
+    This implementation does not match with standards due to various
+    problems with the algorithm according to https://github.com/whatwg/mimesniff/issues/70.
+
+    The current implementation follows
+    https://dxr.mozilla.org/mozilla-central/source/toolkit/components/mediasniffer/mp3sniff.c
+    as the algorithm for MP3 without ID3 sniffing mentioned in standards is originally
+    based on mp3sniff.c.
+    """
+    input_size = len(input_bytes)
+    index = 0
+
+    if not match_mp3_header(input_bytes, input_size, index):
+        return False
+
+    version, bit_rate, freq, pad = parse_mp3_frame(input_bytes)
+
+    skipped_bytes = mp3_framesize(version, bit_rate, freq, pad)
+
+    if skipped_bytes < 4 or skipped_bytes + 4 >= input_size:
+        return False
+
+    index += skipped_bytes
+
+    if match_mp3_header(input_bytes, input_size, index):
+        return True
+    else:
+        return False
+
+
+def is_image(input_bytes: bytes) -> Optional[bytes]:
```

Comment:
As with `is_mp3_non_ID3_signature`, I would expect functions starting with `is_` to be boolean predicates. Could you comment on this, maybe adding docstrings to this and the following functions?

Reply:
Or/and rename them.

Reply:
Right, I meant commenting here on the naming decision, not as in "adding comments" :sweat_smile:

Reply:
I have already changed the names of the functions in other PR. Should I make those changes here too?

Reply:
Right. I guess it’s better to merge as-is, then, otherwise things can get messy when merging.